### PR TITLE
Show jserror indicator in dev mode

### PR DESCRIPTION
**Pull request by lkwg82**
### Purpose

This shows in case of a js warning or error a yellow bar at the top (only in development versions). This dev tool improvement aims to increase maintenance (early bug detection).
Error must log into ```console.error``` or ```console.warn```.

(Besides I added a fix for a 'undefined' error)

(Disclaimer: this is not intended to win design awards. ;))

### Testing

Run a version which has 'dirty' in its version string and provoke a js error. 

### Screenshots

logbar_when_errors

### CODE REVIEW

File: gui/default/index.html [hunk 1]
```diff
@@ -25,6 +25,8 @@
 </head>
 
 <body>
+  <script src="/syncthing/development/logbar.js"></script>
```
Comment (calmh) [inline gui/default/index.html]: I think it's missing this file

File: gui/default/syncthing/core/syncthingController.js [hunk 2]
```diff
@@ -97,6 +97,7 @@ angular.module('syncthing.core')
                 }
 
                 $scope.version = data;
+                $scope.version.isDevelopmentVersion = data.version.indexOf('dirty')>0;
```
Comment (calmh) [inline gui/default/syncthing/core/syncthingController.js]: This is too strict. I think a check for the presence of a "-" in the version would be enough.

**Comment by calmh** [comment]
I dunno... Do we catch all javascript error where needed? For example, what happens when syncthing restarts? With authentication set? Do we really want to show this for everyone who might be running a dev version for whatever reason, without doing any development on the JS side of things?

File: gui/default/index.html (hunk 1 above)
Comment (lkwg82) [inline gui/default/index.html]: ups

**Comment by lkwg82** [comment]
It supports explorative testing, when you click around. I'm not having always the developer console open.

**Comment by calmh** [comment]
To be clear, I meant "does this bring up the ugly error bar in the following common situations that are not JS errors". :)

**Comment by lkwg82** [comment]
Yes, when it is logged through e.g.

```js
console.error('connection error')
```

File: gui/default/index.html (hunk 1 above)
Comment (canton7) [inline gui/default/index.html]: Can we not include scripts with an absolute leading path? If someone serves syncthing from a sub-path, these will break.

**Comment by lkwg82** [comment]
It will also add indicators for missing translation as they are logged with ```warn```.

Maybe this can be switched on, when used with STGUIASSETS variable?

**Comment by calmh** [comment]
Possibly yes. We're going to get a bug report whenever a non-JS-developer sees this, probably even from people experimenting with overriding the CSS and similar, which is why I'm a bit wary of it.

**Comment by lkwg82** [comment]
I see and understand. I'm wary of it too ... . Maybe seeing same amounts of pros and cons disqualifies it.

**Comment by calmh** [comment]
I was going to suggest input from other team members to weigh the issue. But leave this cooking somewhere... We have the deadlock detector (for example) that runs on dev builds but not production, and I could certainly imagine having this as well, as long as we're reasonably convinced that it doesn't pop up for "irrelevant" errors and scares people. :)

**Comment by AudriusButkevicius** [comment]
So I think this has value, as long as we ensure this doesn't pop up in prod builds.

File: gui/default/index.html (hunk 1 above)
Comment (lkwg82) [inline gui/default/index.html]: You are right, will fix that. Missed your comment.

**Comment by lkwg82** [comment]
@AudriusButkevicius Is this detection safe enough?

https://github.com/syncthing/syncthing/pull/2835/commits/bfca57b032333d0805091512bb01cf660d2b8868

**Comment by AudriusButkevicius** [comment]
Seems so.

**Comment by calmh** [comment]
Absolute paths still require fixing, otherwise probably OK. I need to test it a bit, haven't done so yet.

File: gui/default/syncthing/core/syncthingController.js [hunk 3]
```diff
@@ -810,7 +811,7 @@ angular.module('syncthing.core')
         };
 
         $scope.deviceName = function (deviceCfg) {
-            if (typeof deviceCfg === 'undefined') {
+            if (typeof deviceCfg === 'undefined' || deviceCfg.deviceID === undefined) {
```
Comment (calmh) [inline gui/default/syncthing/core/syncthingController.js]: `/me puts on his nit picking hat`

Can we at least check for undefinedness in the same way twice in the same `if`? :)

**Comment by calmh** [comment]
Apart from that, as it irks me, good to go as far as I'm concerned.

File: gui/default/syncthing/core/syncthingController.js (hunk 3 above)
Comment (lkwg82) [inline gui/default/syncthing/core/syncthingController.js]: Sorry for that ... ROFL, just wondered some two minutes what your issue is. :dancers:

**Comment by calmh** [comment]
There are some weird things here. When loading the GUI straight up, I get warnings about every string being missing, although they are in fact not (using the default language). It doesn't show up in the top bar, I guess maybe because the warnings happen prior to the wiring up of the top bar things.

screen shot 2016-03-27 at 09 03 15

If I switch to another language, I get only the appropriate warnings and the top bar does show up.

screen shot 2016-03-27 at 09 05 26

Switching back to English explicitly I get no warnings, so that's fine.

So I guess my concern would primarily be the initial flurry of warnings - maybe those are legit and we're missing something when wiring up the default language?

**Comment by calmh** [comment]
I guess this is mergeable now with the removal of the missing translation warnings?

**Comment by lkwg82** [comment]
I just rescanned and remembered I added a configuration config I missed recently. It is cleaned up now.

**Comment by calmh** [comment]
Github fails to merge this one. Lovely.

**Comment by calmh** [comment]
https://nym.se/r/6gwv.gif

**Comment by calmh** [comment]
(I've asked Github support to take a look at this one, so lets leave it alone for the moment rather than forcing it in.)

**Comment by calmh** [comment]
@st-review merge this please

gui: Show Javascript error indicator on dev builds, remove logging of missing translations

**Comment by st-review** [comment]
OK, merged as 492e92d65d8eea8b60b76d5d082cd17394fb3147. Thanks, @lkwg82!

**Pull request closed without merging**